src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/SingleStringSearchValuesThreeChars.cs

```diff
@@ -0,0 +1,354 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using static System.Buffers.StringSearchValuesHelper;
+
+namespace System.Buffers
+{
+    // Based on SpanHelpers.IndexOf(ref char, int, ref char, int)
+    // This implementation uses 3 precomputed anchor points when searching.
+    // This implementation may also be used for length=2 values, in which case two anchors point at the same position.
+    internal sealed class SingleStringSearchValuesThreeChars<TCaseSensitivity> : StringSearchValuesBase
```

Comment:
How much better is this than just using `IndexOf(string)`?  Is the gain primarily in reduced startup overhead, or does it meaningfully increase throughput because there are fewer false positives on the inner loop?

Reply:
It's a bit of both.

Per-call overhead difference (long text that starts with "the")
|               Method | Word |     Mean |     Error |
|--------------------- |----- |---------:|----------:|
|              IndexOf |  the | 7.465 ns | 0.0811 ns |
|           IndexOfAny |  the | 4.592 ns | 0.0379 ns |
|    IndexOfIgnoreCase |  the | 9.569 ns | 0.1151 ns |
| IndexOfAnyIgnoreCase |  the | 4.582 ns | 0.0177 ns |

An example that highlights the lower false-positive rate: `"twice"` is a word that only appears 16 times in the text, but hits 3889 false positives (4161 under IgnoreCase) with the regular `IndexOf` (looks for `t***e`).
With `SingleStringSearchValuesThreeChars`, it hits 0 false positives (looks for `tw*c*`).

|                       Method |  Word |     Mean |
|----------------------------- |------ |---------:|
|                IndexOf_Count | twice | 88.63 us |
|           SearchValues_Count | twice | 44.51 us |
|      IndexOf_CountIgnoreCase | twice | 89.48 us |
| SearchValues_CountIgnoreCase | twice | 53.43 us |

A few examples similar to #85437
|                       Method |           Value |     Mean |
|----------------------------- |---------------- |---------:|
|                IndexOf_Count |             The | 59.45 us |
|           SearchValues_Count |             The | 51.57 us |
|                              |                 |          |
|                IndexOf_Count |             the | 190.9 us |
|           SearchValues_Count |             the | 148.4 us |
|      IndexOf_CountIgnoreCase |             the | 229.0 us |
| SearchValues_CountIgnoreCase |             the | 158.6 us |
|                              |                 |          |
|                IndexOf_Count |      elementary | 51.77 us |
|           SearchValues_Count |      elementary | 44.03 us |
|      IndexOf_CountIgnoreCase |      elementary | 57.31 us |
| SearchValues_CountIgnoreCase |      elementary | 54.34 us |
|                              |                 |          |
|                IndexOf_Count |          holmes | 57.64 us |
|           SearchValues_Count |          holmes | 44.90 us |
|      IndexOf_CountIgnoreCase |          holmes | 73.23 us |
| SearchValues_CountIgnoreCase |          holmes | 61.10 us |
|                              |                 |          |
|                IndexOf_Count |          watson | 47.81 us |
|           SearchValues_Count |          watson | 45.16 us |
|      IndexOf_CountIgnoreCase |          watson | 53.69 us |
| SearchValues_CountIgnoreCase |          watson | 54.72 us |
|                              |                 |          |
|                IndexOf_Count | sherlock holmes | 78.00 us |
|           SearchValues_Count | sherlock holmes | 44.62 us |
|                              |                 |          |
|                IndexOf_Count | Sherlock Holmes | 44.29 us |
|           SearchValues_Count | Sherlock Holmes | 45.64 us |
|      IndexOf_CountIgnoreCase | Sherlock Holmes | 79.87 us |
| SearchValues_CountIgnoreCase | Sherlock Holmes | 55.01 us |

Same as above, but for top 1000 values
|                       Method |     Mean |    Error |
|----------------------------- |---------:|---------:|
|           Count_SearchValues | 48.88 ms | 0.458 ms |
| CountIgnoreCase_SearchValues | 59.55 ms | 0.261 ms |
|                Count_IndexOf | 53.83 ms | 0.664 ms |
|      CountIgnoreCase_IndexOf | 65.95 ms | 0.345 ms |

Reply:
That's helpful data, thanks.  Based on that then, we should update Regex's "find the next starting position" logic to use this instead of a direct call to IndexOf("..."); I can do so in my PR that adapts Regex to this.